### PR TITLE
Legacy widget rendering endpoint

### PR DESCRIPTION
**Pull request by adamziel**
Related to https://github.com/WordPress/gutenberg/issues/33540
Trac Ticket: https://core.trac.wordpress.org/ticket/53957

This adds a `/wp/v2/widget-types/<widget type>/render` endpoint which accepts POST data and returns the same output as the previous iframe-oriented GET endpoint.

Test plan:

1. Go to `/wp-admin/widgets.php`
2. Paste the following script into your devtools:

```js
wp.apiFetch({
      "path": "/index.php?rest_route=%2Fwp%2Fv2%2Fwidget-types%2Ftext%2Frender&context=edit&per_page=100&_locale=user",
      "method": "POST",
      "data": {"id_base":"text","instance":{"encoded":"YTo0OntzOjU6InRpdGxlIjtzOjU6InRlc3QyIjtzOjQ6InRleHQiO3M6NToidGVzdDMiO3M6NjoiZmlsdGVyIjtiOjE7czo2OiJ2aXN1YWwiO2I6MTt9","hash":"6bd53d501ad76cfbc50ef417d78dc9f1","raw":{"title":"test2","text":"test3","filter":true,"visual":true}}}
})
.then(console.log.bind(console))
.catch(console.error.bind(console));
```

3. Confirm the response was successful and the widget got rendered as expected.

**Alternatives considered:**
* `ServerSideRender` – we need to render a full HTML document with `wp_head()`, CSS links, maybe some additional scripts, and this is much more that I would like to return from a block rendering API.
* Updating the existing `admin_init` hook handler to work with `POST` requests – it's tricky and doesn't seem right, I don't think we should stretch hooks like that. This clearly is an API request so let's treat it as such.

### CODE REVIEW

**Comment by adamziel** [comment]
All the tests pass! @TimothyBJacobs – do you fancy a review?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (anton-vlasenko) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:427]:
```suggestion
				isset($request['instance']) ? $request['instance'] : null
```
As far as I understand `instance` request parameter doesn't have a default value, so I would check if it is present in the `$request` array before using it.
The `instance` array key could be undefined, and this can generate a PHP warning.
Alternatively, we could define a default value for the `instance` parameter here: https://github.com/WordPress/wordpress-develop/blob/b0b6cb921792254fd9ba4d93a2263c17638d0670/src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php#L118

**Review by anton-vlasenko — CHANGES_REQUESTED** [review]
Good job!
I only have 1 small remark.

**Comment by adamziel** [comment]
Great suggestion @anton-vlasenko ! I just added it.

**Review by anton-vlasenko — APPROVED** [review]
Everything looks good 👍

**Comment by draganescu** [comment]
I think we should bring this in along with the change in the legacy widget that uses it.

**Comment by anton-vlasenko** [comment]
> I think we should bring this in along with the change in the legacy widget that uses it.

@draganescu 
I thought I need to wait until this PR is merged (I mean not this PR, but the corresponding Trac issue).
Sure, I will modify my PR to make use of that new endpoint.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (noisysocks) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:425]: Would it make sense to return a response with `Content-type: text/html` rather than have all of the HTML inside this JSON string?

**Comment by noisysocks** [comment]
How will the iframe use this request? Can iframes make `POST` requests?

I'd also like to get the @TimothyBJacobs tick of approval before merging this 🙂

**Comment by kevin940726** [comment]
@noisysocks I think the idea is to make that request using `apiFetch` and get the response as text then just feed it into the `srcdoc` prop of the iframe.

```jsx
const response = await apiFetch({ path: '...' });
const html = await response.text();

<iframe srcDoc={html} />
```

In that case, I think this approach is very doable and could be backward-compatible and minimal.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (adamziel) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:425]: I considered that for a moment and decided against it. The reason is that:
* In success case we get an HTML response with `Content-type: text/html` which means we need to set `parse: false` in the `apiFetch`
* But in the error case we would still get JSON, so `parse: false` means we need to parse manually and I don't like that

If you have some ideas how to make the JS part better, I'm all ears!

**Comment by adamziel** [comment]
@noisysocks exactly as @kevin940726 said. We may either set `srcDoc`, or tamper with `contentDocument`. Either way, the iframe would be only a vessel for the data we get via `apiFetch`.

**Comment by spacedmonkey** [comment]
CC @TimothyBJacobs

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:441]: Why is the permission check here instead of in the `permission_callback`? This should go into its own `permission_callback` specific for this endpoint, that can then call `check_read_permission`.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:423]: This needs a docblock.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:434]: Missing `@since` tag.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:425]: Yeah let's stick with this format, it matches what we do in the block renderer endpoint.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:445]: This probably needs a `defined()` check.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:485]:
```suggestion
		return ob_get_clean();
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php:441]: We should actually probably just reuse `get_item_permission_callback` which matches what we do in the encode endpoint.

**Comment by adamziel** [comment]
@TimothyBJacobs thank you for reviewing! I just addressed all your feedback.

**Comment by TimothyBJacobs** [comment]
Thanks @adamziel! The only small tweak is that the `@since` tags [should come before](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods) the `@param` tags.

**Comment by noisysocks** [comment]
Last minute feedback, sorry!

Did you consider incorporating this functionality into the existing `/widget-types/:id/encode` endpoint? `/encode` is an existing RPC-style endpoint that already accepts `instance` as an argument and returns `preview` in the response. The only difference is that `/encode` does not output all of the wrapping HTML including scripts and styles and whatnot. Perhaps it could, though.

**Comment by noisysocks** [comment]
> Now that I think about it, maybe we should only add it to Gutenberg repo and not here 🤔 and then backport it on the next major release

Yes we probably should do this. That way it can be used by https://github.com/WordPress/gutenberg/pull/34184, too.

**Comment by adamziel** [comment]
Let's continue in the Gutenberg repo then: https://github.com/WordPress/gutenberg/pull/34230

**Pull request closed without merging**
